src/main/scala/viper/gobra/reporting/StatsCollector.scala

```diff
@@ -0,0 +1,376 @@
+    viperMemberNameGobraMemberMap.get(viperMemberKey(taskName, viperMemberName, pkgDir))
+      .orElse(viperMemberNameGobraMemberMap.get(viperMemberKey(taskName, viperMemberName)))
+
+  def getWarnings(pkgDir: String, pkg: String, config: Config): Set[String] =
+    memberMap.values
+      .filter(gobraMember => gobraMember.pkgDir == pkgDir && gobraMember.pkg == pkg)
+      .flatMap(gobraMember => {
+        val name = gobraMember.pkg + "." + gobraMember.memberName + gobraMember.args
+        gobraMember.dependencies().flatMap({
+          // Trusted implies abstracted, so we match trusted first
+          case GobraMemberEntry(_, pkg, memberName, _, args, _, true, _) =>
+            Some("Warning: Member " + name + " depends on trusted member " + pkg + "." + memberName + args + "\n")
+          case GobraMemberEntry(_, pkg, memberName, _, args, _, _, true) =>
+            Some("Warning: Member " + name + " depends on abstract member " + pkg + "." + memberName + args + "\n")
+          case GobraMemberEntry(pkgDir, pkg, _, _, _, _, _, _) if !config.inputPackageMap.contains(PackageEntry(pkgDir, pkg)) =>
+            Some("Warning: Depending on imported package that is not verified: " + pkgDir + " - " + pkg)
+          case _ => None
+        })
+      }).toSet
+
+  def addResult(pkgDir: String,
+                pkg: String,
+                memberName: String,
+                args: String,
+                nodeType: GobraNodeType,
+                viperMember: ViperMemberEntry,
+                isTrusted: Boolean,
+                isAbstract: Boolean): Unit = {
+    val key = gobraMemberKey(pkgDir, pkg, memberName, args)
+
+    val viperKey = viperMemberKey(viperMember.taskName, viperMember.memberName, pkgDir)
+    val shortViperKey = viperMemberKey(viperMember.taskName, viperMember.memberName)
+
+    memberMap.get(key) match {
+      case Some(existing) =>
+        existing.viperMembers.get(viperKey).orElse(existing.viperMembers.get(shortViperKey)) match {
+          case Some(existingViperEntry) =>
+            // Merge Viper members that occur multiple times during a task(used for chopper)
+            existing.viperMembers.put(viperKey, ViperMemberEntry(
+              existingViperEntry.memberName,
+              existingViperEntry.taskName,
+              existingViperEntry.time + viperMember.time,
+              existingViperEntry.nodeType,
+              existingViperEntry.dependencies ++ viperMember.dependencies,
+              existingViperEntry.success && viperMember.success,
+              existingViperEntry.cached || viperMember.cached,
+              existingViperEntry.fromImport,
+              existingViperEntry.hasBody || viperMember.hasBody
+            ))
+          case None => existing.viperMembers.put(viperKey, viperMember)
+        }
+
+        // If we encounter an abstract version of a member, we know for sure, its abstract
+        val newAbstract = existing.isAbstract || isAbstract
```

Comment:
why is disjunction used here? Wouldn't a member occur once non-abstract and possibly multiple times abstract in all chopped parts?

Reply:
Note that this abstract flag is for Gobra members and not viper members. Since we don't check if imported members are abstract, an abstract gobra member will only occur once, when the package it belongs to is verified

Reply:
The viper members have no influence on this, it is produced from the `PNode` stored in the `info` field on a message

We prevent the second case from happening by defining the `isAbstract` field this way: 
` isAbstract = p.body.isEmpty && !isImported`
where `p` is a `PNode` with a body
This way we prevent false positives from imported packages, the downside is that we can't say if a member is abstract if he only appears as a imported member.

Reply:
Thanks. I missed this part. Maybe renaming `isAbstract` to `isAbstractAndNotImported` would help to make the code more readable